File: history/day_11.ts
import { Logger, Part, run, Type } from "../day_utils"

interface StarPos {
    orig_index: number,
    x: number,
    y: number
}

function parse(lines: string[]): StarPos[] {
    let count = 1;
    return lines.flatMap((l, y) => l.split("")
        .mapNonNull((c, x) => {
            if (c !== "#") {
                return undefined;
            }
            const star: StarPos = { x, y, orig_index: count };
            count++;
            return star;
        })
    )
}

function buildOffsetMap(max: number, non_empty: Set<number>, part: Part, type: Type): Map<number, number> {
    const result = new Map<number, number>();
    let offset = 0;
    for (let pos = 0; pos <= max; ++pos) {
        result.set(pos, offset);
        if (!non_empty.has(pos)) {
            if (part === Part.PART_1) {
                offset++;
            } else {
                offset += (type === Type.TEST ? 100 : 1_000_000) - 1;
            }
        }
    }
    return result;
}

function solve(stars: StarPos[], part: Part, type: Type): bigint {
    const non_empty_x = new Set<number>();
    const non_empty_y = new Set<number>();
    const { maxX, maxY } = stars.reduce((maxInfo, star) => {
        non_empty_x.add(star.x);
        non_empty_y.add(star.y);
        return {
            maxX: Math.max(star.x, maxInfo.maxX),
            maxY: Math.max(star.y, maxInfo.maxY)
        }
    }, { maxX: 0, maxY: 0 });

    const offsetXMap = buildOffsetMap(maxX, non_empty_x, part, type);
    const offsetYMap = buildOffsetMap(maxY, non_empty_y, part, type);

    const correctedStars = stars.map(s => {
        return { x: s.x + offsetXMap.get(s.x)!, y: s.y + offsetYMap.get(s.y)! }
    });
    return correctedStars.reduce(
        (total, s1, index) => {
            return correctedStars.slice(index + 1).map(s2 => {
                return Math.abs(s2.x - s1.x) + Math.abs(s2.y - s1.y);
            }).reduce((total_pair, d_pair) => total_pair + BigInt(d_pair), total)
        }
        , 0n)
}


function puzzle(lines: string[], part: Part, type: Type, logger: Logger): void {
    const data = parse(lines);
    const result = solve(data, part, type);
    if (part === Part.PART_1) {
        logger.result(result, [374n, 9693756n])
    }
    else {
        logger.result(result, [8410n, 717878258016n])
    }
}

/**
 * Update the date number after copy
 * Adapt types list to your needs and parts also 
 */
run(11, [Type.TEST, Type.RUN], puzzle, [Part.PART_1, Part.PART_2])

File: rust/src/day11.rs
use std::collections::{BTreeSet, BTreeMap};

use crate::{
    check_result,
    utils::{Context, Part},
};

struct Star{
    x:usize,
    y:usize,
}

fn build_offset_map(max: usize, non_empty: &BTreeSet<usize>, context:&Context)-> BTreeMap<usize,usize> {
    let mut result = BTreeMap::new();
    let mut offset = 0;
    for  pos in 0..=max {
        result.insert(pos, offset);
        if !non_empty.contains(&pos) {
            if context.is_part(Part::Part1) {
                offset+=1;
            } else {
                offset += if context.is_test() { 100 } else { 1_000_000 } - 1;
            }
        }
    }
    return result;
}

fn solve(lines: &Vec<String>,context: &Context) -> usize {
    let mut used:(BTreeSet<usize>,BTreeSet<usize>) = (BTreeSet::new(),BTreeSet::new());
    let mut max:(usize,usize) = (0,0);
    let stars= lines
        .iter()
        .enumerate()
        .flat_map(|(y,l)|{
            l.chars().enumerate().filter_map(|(x,c)|
                if c=='.' {
                    None
                } else {
                    used.0.insert(x);
                    used.1.insert(y);
                    max.0=max.0.max(x);
                    max.1=max.1.max(y);
                    Some(Star{x,y})
                }
            ).collect::<Vec<Star>>()
        })
        .collect::<Vec<Star>>();
    let offset_x_map = build_offset_map(max.0, &used.0,  context);
    let offset_y_map = build_offset_map(max.1, &used.1,  context);

    let corrected_stars = stars.iter().map(|s| {
        return Star{ x: s.x + offset_x_map.get(&s.x).unwrap(), y: s.y + offset_y_map.get(&s.y).unwrap() }
    }).collect::<Vec<Star>>();

    return corrected_stars.iter().enumerate().fold(0,
        |total, (index,s1)| {
            total + corrected_stars.iter().skip(index).map(|s2| s2.x.abs_diff(s1.x) + s2.y.abs_diff(s1.y)).sum::<usize>()
        })
}

pub fn puzzle(context: &Context, lines: &Vec<String>) {
    let result = solve(lines,context);
    if context.is_part(Part::Part1) {
        check_result!(context, result, [374, 9693756]);
    } else {
        check_result!(context, result, [8410, 717878258016]);
    }
}
